File: src/main/java/dev/stocky37/epic7/json/Hero.java
package dev.stocky37.epic7.json;

import com.google.common.base.Converter;
import dev.stocky37.epic7.core.Stat;
import dev.stocky37.epic7.core.Stats;

import javax.json.Json;
import javax.json.JsonObject;
import javax.json.JsonObjectBuilder;
import javax.json.JsonValue;
import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

public class Hero extends ForwardingJsonObject {
	private static final Converter<Stats, JsonObject> converter = StatsJsonConverter.instance();
	private final JsonObject json;

	public Hero(JsonObject json) {
		this.json = json;
	}

	private static JsonObject combineAwakeningStats(Iterable<JsonObject> arr) {
		final JsonObjectBuilder builder = Json.createObjectBuilder();
		for(final JsonObject stats : arr) {
			stats.forEach((key, value) -> {
				final Stat stat = Stat.fromId(key);
				final BigDecimal numValue = new BigDecimal(value.toString());
				// if it's NOT a %stat and it's value is less than 1, assume it's meant to be a percent of the base stat
				final String newKey = !stat.isPercentage() && numValue.compareTo(BigDecimal.ONE) < 0 ? key + "%" : key;
				builder.add(newKey, numValue);
			});
		}
		return builder.build();
	}

	@Override
	protected JsonObject delegate() {
		return json;
	}

	public Stats getLvl50Stats() {
		return converter.reverse().convert(getStatsJson().getJsonObject("lv50FiveStarNoAwaken"));
	}

	public Stats getLvl60Stats() {
		return converter.reverse().convert(getStatsJson().getJsonObject("lv60SixStarNoAwaken"));
	}

	public Iterable<Stats> getAwakeningStats() {
		return converter.reverse().convertAll(getAwakeningStatsJson());
	}

	private List<JsonObject> getAwakeningStatsJson() {
		return getAwakeningJson().stream()
			.map(it -> it.getJsonArray("statsIncrease").getValuesAs(JsonValue::asJsonObject))
			.map(Hero::combineAwakeningStats)
			.collect(Collectors.toList());
	}

	private List<JsonObject> getAwakeningJson() {
		return json.getJsonArray("awakening").getValuesAs(JsonValue::asJsonObject);
	}

	private JsonObject getStatsJson() {
		return json.getJsonObject("stats");
	}


}

File: src/main/kotlin/dev/stocky37/epic7/json/Hero.kt
package dev.stocky37.epic7.json

import dev.stocky37.epic7.core.Stat
import dev.stocky37.epic7.core.Stats
import java.math.BigDecimal
import javax.json.Json
import javax.json.JsonObject

data class Hero(private val json: JsonObject) : JsonObject by json {
	val lv50Stats: Stats by lazy {
		getStats(statsJson.getJsonObject("lv50FiveStarNoAwaken"))
	}

	val lv60Stats: Stats by lazy {
		getStats(statsJson.getJsonObject("lv60SixStarNoAwaken"))
	}

	val awakenings: List<Stats> by lazy {
		awakeningsStatsJson.map { getStats(it) }
	}

	private val statsJson: JsonObject by lazy {
		json.getJsonObject("stats")
	}


	private val awakeningsJson: List<JsonObject> by lazy {
		json.getJsonArray("awakening").getValuesAs(JsonObject::class.java)
	}

	private val awakeningsStatsJson: List<JsonObject> by lazy {
		awakeningsJson
			.map { it.getJsonArray("statsIncrease").getValuesAs(JsonObject::class.java) }
			.map { combineAwakeningStats(it) }
	}

//	private val awakeningStatPairs: List<List<Pair<Stat, BigDecimal>>> by lazy {
//		awakeningsStatsJson.map { it.map { json -> json. } }
//	}

//	fun getPairs(json: JsonObject): List<Pair<Stat, BigDecimal>> {
//		return json.map { (key, node) -> {
//			val baseStat = Stat.fromId(key)
//			val value = BigDecimal(node.toString())
//			val stat = if(value < BigDecimal.ONE) Stat.fromId("$key%") else baseStat
//			Pair(stat, value)
//		} }
//	}


	companion object {
		private fun getStats(json: JsonObject): Stats {
			return Stats(json
				.filterKeys { !it.equals("cp", true) }
				.mapKeys { Stat.fromId(it.key) }
				.mapValues { BigDecimal(it.value.toString()) }
			)
		}

		private fun combineAwakeningStats(jsonArray: List<JsonObject>): JsonObject {
			val builder = Json.createObjectBuilder()
			jsonArray.forEach {
				it.forEach { (key, value) ->
					run {
						// handle the fact that
						val newKey = if(
							!Stat.fromId(key).isPercentage
							&& BigDecimal(value.toString()) < BigDecimal.ONE
						) "$key%" else key
						builder.add(newKey, value)
					}
				}
			}
			return builder.build()
		}
	}
}